File: Protocol/_theoreme_sim/src/phases/upRoundPhase.js
/* Purpose: Simulate one UP phase sweep where each device attempts to send toward gateway. */

import { THEOREM_ROOT_ID } from "../core/constants.js";
import {
  getLinkDeliveryProbability,
  recordLinkUsage,
} from "../propagation/linkUsageTracker.js";

/**
 * @param {number} observerId
 * @param {number} senderId
 * @param {any} state
 * @returns {boolean}
 */
function updateNeighborEstimate(observerId, senderId, state) {
  const observer = state.nodes.get(observerId);
  const sender = state.nodes.get(senderId);
  const estimateMap = state.estimates.get(observerId);

  if (!observer || !sender || !estimateMap) {
    return false;
  }

  const previous = estimateMap.get(senderId) ?? 0;
  const next = Math.max(previous, sender.qTotal);
  if (next > previous) {
    estimateMap.set(senderId, next);
    return true;
  }

  return false;
}

/**
 * @param {number} nodeId
 * @param {Set<number>} visited
 * @param {any} state
 * @returns {number|null}
 */
function chooseUpNextHop(nodeId, visited, state) {
  const node = state.nodes.get(nodeId);
  if (!node) {
    return null;
  }

  const preferredParent = state.parentMap.get(nodeId) ?? node.parent ?? null;
  if (preferredParent !== null && !visited.has(preferredParent)) {
    const parent = state.nodes.get(preferredParent);
    if (parent && parent.qTotal > node.qTotal) {
      return preferredParent;
    }
  }

  const neighbors = state.adjacency.get(nodeId) ?? new Set();
  let bestId = null;
  let bestCharge = Number.NEGATIVE_INFINITY;

  for (const neighborId of neighbors.values()) {
    if (visited.has(neighborId)) {
      continue;
    }

    const neighbor = state.nodes.get(neighborId);
    if (!neighbor) {
      continue;
    }

    if (neighbor.qTotal <= node.qTotal) {
      continue;
    }

    if (
      neighbor.qTotal > bestCharge ||
      (neighbor.qTotal === bestCharge &&
        (bestId === null || neighborId < bestId))
    ) {
      bestCharge = neighbor.qTotal;
      bestId = neighborId;
    }
  }

  return bestId;
}

/**
 * @param {any} state
 * @returns {{attempted:number,reachedGateway:number,hops:number,updates:number}}
 */
export function runUpRoundPhase(state) {
  const hopGain = Math.max(
    1,
    Math.round(Math.max(1, state.config.chargeDropPerHop) * 0.03),
  );

  let attempted = 0;
  let reachedGateway = 0;
  let hops = 0;
  let updates = 0;

  const nodeIds = [...state.nodes.keys()].sort((a, b) => a - b);

  for (const sourceId of nodeIds) {
    if (sourceId === THEOREM_ROOT_ID) {
      continue;
    }

    attempted += 1;

    let currentId = sourceId;
    const visited = new Set([sourceId]);
    const safetyLimit = Math.max(1, state.nodes.size);

    for (let step = 0; step < safetyLimit; step += 1) {
      const nextId = chooseUpNextHop(currentId, visited, state);
      if (nextId === null) {
        break;
      }

      const sender = state.nodes.get(currentId);
      const receiver = state.nodes.get(nextId);
      if (!sender || !receiver) {
        break;
      }

      const probability = getLinkDeliveryProbability(state, currentId, nextId);
      if (state.rng() > probability) {
        break;
      }

      recordLinkUsage(state, currentId, nextId, 1);

      const senderBefore = sender.qTotal;
      const receiverBefore = receiver.qTotal;

      sender.qTotal += hopGain;
      receiver.qTotal += hopGain;

      if (sender.qTotal > senderBefore || receiver.qTotal > receiverBefore) {
        updates += 1;
      }

      if (updateNeighborEstimate(nextId, currentId, state)) {
        updates += 1;
      }

      hops += 1;
      currentId = nextId;

      if (currentId === THEOREM_ROOT_ID) {
        reachedGateway += 1;
        break;
      }

      if (visited.has(currentId)) {
        break;
      }
      visited.add(currentId);
    }
  }

  return {
    attempted,
    reachedGateway,
    hops,
    updates,
  };
}

File: Protocol/_theoreme_sim_py/src/phases/up_round_phase.py
"""Purpose: Simulate one UP phase where devices try to route greedily toward gateway."""

from __future__ import annotations

from src.core.constants import THEOREM_ROOT_ID
from src.propagation.link_usage_tracker import (
    get_link_delivery_probability,
    record_link_usage,
)
from src.utils.math_utils import js_round


def _update_neighbor_estimate(observer_id: int, sender_id: int, state: dict) -> bool:
    observer = state["nodes"].get(observer_id)
    sender = state["nodes"].get(sender_id)
    estimate_map = state["estimates"].get(observer_id)
    if observer is None or sender is None or estimate_map is None:
        return False

    previous = float(estimate_map.get(sender_id, 0.0))
    nxt = max(previous, float(sender.q_total))
    if nxt > previous:
        estimate_map[sender_id] = nxt
        return True
    return False


def _choose_up_next_hop(node_id: int, visited: set[int], state: dict) -> int | None:
    node = state["nodes"].get(node_id)
    if node is None:
        return None

    preferred_parent = state["parentMap"].get(node_id)
    if preferred_parent is None:
        preferred_parent = node.parent

    if preferred_parent is not None and preferred_parent not in visited:
        parent = state["nodes"].get(preferred_parent)
        if parent and float(parent.q_total) > float(node.q_total):
            return preferred_parent

    neighbors = state["adjacency"].get(node_id, set())
    best_id: int | None = None
    best_charge = float("-inf")

    for neighbor_id in sorted(neighbors):
        if neighbor_id in visited:
            continue
        neighbor = state["nodes"].get(neighbor_id)
        if neighbor is None:
            continue
        if float(neighbor.q_total) <= float(node.q_total):
            continue

        if (
            float(neighbor.q_total) > best_charge
            or (
                float(neighbor.q_total) == best_charge
                and (best_id is None or neighbor_id < best_id)
            )
        ):
            best_charge = float(neighbor.q_total)
            best_id = neighbor_id

    return best_id


def run_up_round_phase(state: dict) -> dict:
    hop_gain = max(
        1,
        js_round(max(1.0, float(state["config"]["chargeDropPerHop"])) * 0.03),
    )

    attempted = 0
    reached_gateway = 0
    hops = 0
    updates = 0

    node_ids = sorted(state["nodes"].keys())
    for source_id in node_ids:
        if source_id == THEOREM_ROOT_ID:
            continue

        attempted += 1
        current_id = source_id
        visited = {source_id}
        safety_limit = max(1, len(state["nodes"]))

        for _ in range(safety_limit):
            next_id = _choose_up_next_hop(current_id, visited, state)
            if next_id is None:
                break

            sender = state["nodes"].get(current_id)
            receiver = state["nodes"].get(next_id)
            if sender is None or receiver is None:
                break

            probability = get_link_delivery_probability(state, current_id, next_id)
            if state["rng"]() > probability:
                break

            record_link_usage(state, current_id, next_id, 1)

            sender_before = float(sender.q_total)
            receiver_before = float(receiver.q_total)
            sender.q_total += hop_gain
            receiver.q_total += hop_gain

            if float(sender.q_total) > sender_before or float(receiver.q_total) > receiver_before:
                updates += 1

            if _update_neighbor_estimate(next_id, current_id, state):
                updates += 1

            hops += 1
            current_id = next_id

            if current_id == THEOREM_ROOT_ID:
                reached_gateway += 1
                break

            if current_id in visited:
                break
            visited.add(current_id)

    return {
        "attempted": attempted,
        "reachedGateway": reached_gateway,
        "hops": hops,
        "updates": updates,
    }
